File: app/src/main/java/com/gallosalocin/go4lunch/viewmodels/RestaurantViewModel.java
package com.gallosalocin.go4lunch.viewmodels;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.gallosalocin.go4lunch.models.RestaurantResult;
import com.gallosalocin.go4lunch.services.RestaurantApi;
import com.gallosalocin.go4lunch.services.dto.DetailsResponse;
import com.gallosalocin.go4lunch.services.dto.DetailsResult;
import com.gallosalocin.go4lunch.services.dto.RestaurantResponse;

import org.jetbrains.annotations.NotNull;

import java.util.List;

import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;
import timber.log.Timber;

import static com.gallosalocin.go4lunch.util.Constants.BASE_URL;

public class RestaurantViewModel extends AndroidViewModel {

    private RestaurantApi restaurantApi;
    private MutableLiveData<List<RestaurantResult>> restaurantList;
    private MutableLiveData<DetailsResult> detailsResult;

    public RestaurantViewModel(@NonNull Application application) {
        super(application);
    }

    public LiveData<List<RestaurantResult>> getRestaurants(String currentLocation, int radius, String type, String key) {
        if (restaurantList == null) {
            restaurantList = new MutableLiveData<>();
            loadRestaurants(currentLocation, radius, type, key);
        }
        return restaurantList;
    }

    public LiveData<DetailsResult> getDetailsRestaurant(String placeId, String key) {
        detailsResult = new MutableLiveData<>();
        loadDetailsRestaurant(placeId, key);
        return detailsResult;
    }

    private void setupRestaurantApi() {
        HttpLoggingInterceptor loggingInterceptor = new HttpLoggingInterceptor();
        loggingInterceptor.setLevel(HttpLoggingInterceptor.Level.BODY);

        OkHttpClient okHttpClient = new OkHttpClient().newBuilder()
                .addInterceptor(loggingInterceptor)
                .build();

        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .client(okHttpClient)
                .build();

        restaurantApi = retrofit.create(RestaurantApi.class);
    }

    private void loadRestaurants(String currentLocation, int radius, String type, String key) {
        setupRestaurantApi();

        Call<RestaurantResponse> restaurantResponseCall = restaurantApi.getNearbyRestaurant(currentLocation, radius, type, key);
        restaurantResponseCall.enqueue(new Callback<RestaurantResponse>() {
            @Override
            public void onResponse(@NotNull Call<RestaurantResponse> call, @NotNull Response<RestaurantResponse> response) {
                if (response.body() != null) {
                    restaurantList.setValue(response.body().getRestaurantResults());
                }
            }

            @Override
            public void onFailure(@NotNull Call<RestaurantResponse> call, @NotNull Throwable t) {
            }
        });
    }

    private void loadDetailsRestaurant(String placeId, String key) {
        setupRestaurantApi();

        Call<DetailsResponse> detailsResponseCall = restaurantApi.getDetailsRestaurant(placeId, key);
        detailsResponseCall.enqueue(new Callback<DetailsResponse>() {
            @Override
            public void onResponse(@NotNull Call<DetailsResponse> call, @NotNull Response<DetailsResponse> response) {
                if (response.body() != null) {
                    detailsResult.setValue(response.body().getDetailsResult());
                }
            }

            @Override
            public void onFailure(@NotNull Call<DetailsResponse> call, @NotNull Throwable t) {
                Timber.e("Error loadDetailsRestaurant : %s", t.getMessage());
            }
        });

    }

}

File: app/src/main/java/com/gallosalocin/go4lunch/viewmodels/RestaurantViewModel.kt
package com.gallosalocin.go4lunch.viewmodels

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import com.gallosalocin.go4lunch.data.network.dto.DetailsResponse
import com.gallosalocin.go4lunch.data.network.dto.DetailsResult
import com.gallosalocin.go4lunch.data.network.dto.RestaurantResponse
import com.gallosalocin.go4lunch.data.repositories.Repository
import com.gallosalocin.go4lunch.models.RestaurantResult
import dagger.hilt.android.lifecycle.HiltViewModel
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.Single
import io.reactivex.rxjava3.disposables.Disposable
import io.reactivex.rxjava3.schedulers.Schedulers
import io.reactivex.rxjava3.subjects.PublishSubject
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import timber.log.Timber
import javax.inject.Inject


@HiltViewModel
class RestaurantViewModel @Inject constructor(
        application: Application,
        private var repository: Repository
) : AndroidViewModel(application) {

    private lateinit var disposable: Disposable

    private lateinit var restaurantListPS: PublishSubject<List<RestaurantResult>>

    var detailsResultPS: PublishSubject<DetailsResult> = PublishSubject.create()

    fun getNearbyRestaurantList(currentLocation: String, radius: Int, type: String, key: String): PublishSubject<List<RestaurantResult>> {
        restaurantListPS = PublishSubject.create()
        apiCallNearbyRestaurant(currentLocation, radius, type, key)
        return restaurantListPS
    }

    private fun apiCallNearbyRestaurant(currentLocation: String, radius: Int, type: String, key: String) {
        disposable = Observable.create<List<RestaurantResult>> {
            repository.remote.getNearbyRestaurants(currentLocation, radius, type, key).enqueue(object : Callback<RestaurantResponse> {
                override fun onResponse(call: Call<RestaurantResponse>, response: Response<RestaurantResponse>) {
                    it.onNext(response.body()?.restaurantResults)
                }

                override fun onFailure(call: Call<RestaurantResponse>, throwable: Throwable) {
                    Timber.e("$throwable")
                }
            })
        }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        { restaurantListPS.onNext(it) },
                        { error -> Timber.e("$error") },
                )
    }

    fun apiCallDetailsRestaurant(placeId: String, key: String) {
        disposable = Single.create<DetailsResult> {
            repository.remote.getDetailsRestaurant(placeId, key).enqueue(object : Callback<DetailsResponse?> {
                override fun onResponse(call: Call<DetailsResponse?>, response: Response<DetailsResponse?>) {
                    it.onSuccess(response.body()?.detailsResult)
//                    it.onNext(response.body()?.detailsResult)
                }

                override fun onFailure(call: Call<DetailsResponse?>, throwable: Throwable) {
                    Timber.e("$throwable")
                }
            })
        }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        { detailsResultPS.onNext(it) },
                        { error -> Timber.e("$error") }
                )
    }

    override fun onCleared() {
        super.onCleared()
        if (!disposable.isDisposed) {
            disposable.dispose()
        }
    }
}